ignite/handlers/ema_handler.py

```diff
@@ -0,0 +1,155 @@
+            raise ValueError(f"Invalid momentum_warmup: {momentum_warmup}")
+        if not momentum_warmup <= momentum:
+            raise ValueError(
+                f"momentum_warmup should be less than or equal to momentum, but got "
+                f"momentum_warmup: {momentum_warmup} and momentum: {momentum}"
+            )
+        if not (isinstance(interval, int) and interval > 0):
+            raise ValueError(f"Invalid interval: {interval}")
+        if not isinstance(warmup_iters, int) and warmup_iters > 0:
+            raise ValueError(f"Invalid warmup_iters: {warmup_iters}")
+        if not isinstance(model, nn.Module):
+            raise ValueError(
+                f"model should be an instance of nn.Module or its subclasses, but got"
+                f"model: {model.__class__.__name__}"
+            )
+        self.momentum = momentum
+        self.interval = interval
+        self.momentum_warmup = momentum_warmup
+        self.warmup_iters = warmup_iters
+
+        self.unwrapped_model = self._unwrap_model(model)
+        self.ema_model = deepcopy(self.unwrapped_model)
+        self.ema_model.eval()
+
+    @staticmethod
+    def _unwrap_model(model: nn.Module) -> nn.Module:
+        if isinstance(model, (nn.parallel.DataParallel, nn.parallel.DistributedDataParallel)):
+            return model.module
+        else:
+            return model
+
+    def _get_momentum(self, curr_iter: int) -> float:
+        """Get current momentum, `curr_iter` should be 1-based. When `curr_iter = 1`, `momentum =
+        self.momentum_warmup`; when `curr_iter >= self.warmup_iters`, `momentum = self.momentum`"""
+        if curr_iter < 1:
+            raise ValueError(f"curr_iter should be at least 1, but got {curr_iter}.")
+        denominator = max(1, self.warmup_iters - 1)
+        momentum = self.momentum_warmup + (self.momentum - self.momentum_warmup) * (curr_iter - 1) / denominator
+        return min(self.momentum, momentum)
+
+    @torch.no_grad()
+    def _update_weights(self, engine: Engine) -> None:
+        """Update weights of ema model"""
+        momentum = engine.state.ema_momentum
+
+        for ema_v, model_v in zip(self.ema_model.state_dict().values(), self.unwrapped_model.state_dict().values()):
+            ema_v.mul_(1.0 - momentum).add_(model_v.data, alpha=momentum)  # type: ignore
+
+    def _update_ema_momentum(self, engine: Engine) -> None:
+        """Update momentum in engine.state"""
+        curr_iter = engine.state.iteration
+        momentum = self._get_momentum(curr_iter)
+        engine.state.ema_momentum = momentum
+
+    def attach(self, engine: Engine) -> None:
+        """Attach the handler to engine.
+
+        Args:
+            engine: Trainer to which the handler will be attached.
+
+        """
+        engine.add_event_handler(Events.ITERATION_STARTED, self._update_ema_momentum)
+        engine.add_event_handler(Events.ITERATION_COMPLETED(every=self.interval), self._update_weights)
```

Comment:
Instead of hard-coding the event here as `Events.ITERATION_COMPLETED(every=self.interval)` we can add that as the argument to `attach`:
```python
    def attach(
        self, 
        engine: Engine, 
        event_name: Union[str, Events, CallableEventWithFilter, EventsList] = Events.ITERATION_COMPLETED
    ) -> None:
        """Attach the handler to engine.
        Args:
            engine: Trainer to which the handler will be attached.
        """
        engine.add_event_handler(event_name, self._update_ema_momentum)
        engine.add_event_handler(event_name, self._update_weights)
```

Reply:
There are two issues with this API:

- `_update_ema_momemtum` and `_update_weights` use the same event, e.g., `Events.ITERATION_COMPLETED`. So we don't know which function will be called first. If the trainer first update EMA model, then it will raise an error since `Engine.state.ema_momentum` has not been set yet.
- Consider the case where two EMA handlers are attached to the same trainer. The one handler's momentum will override the other one's, depending on which one's `_update_ema_momentum` is called first. Because `Engine.state.ema_momentum` is a float. 

I tackle these issues in the next update, where I use two separate events `momentum_event` and `model_event`, and initialize `Engine.state.ema_momentum` as a dict.

Reply:
> _update_ema_momemtum and _update_weights use the same event, e.g., Events.ITERATION_COMPLETED. So we don't know which function will be called first. 

We know which function is called the first -> the first added -> `engine.add_event_handler(event_name, self._update_ema_momentum)`

> Consider the case where two EMA handlers are attached to the same trainer. The one handler's momentum will override the other one's, depending on which one's _update_ema_momentum is called first. Because Engine.state.ema_momentum is a float.

yes, this is a good point. I think we can do the following:
```python
    def attach(
        self, 
        engine: Engine, 
        event_name: Union[str, Events, CallableEventWithFilter, EventsList] = Events.ITERATION_COMPLETED,
        name="ema_momentum",
    ) -> None:

        if hasattr(engine.state, name):
            # warn about colliding attribute
        setattr(engine.state, name, 0.0)
        engine.add_event_handler(event_name, self._update_ema_momentum, name)
        engine.add_event_handler(event_name, self._update_weights)
```

Reply:
If we initialize `Engine.state.ema_momentum` as a dict, there is no need to warn. We can have multiple EMA handlers attached to the same engine, as long as the keys for `Engine.state.ema_momentum` are different. For example, one for generator and one of discriminator. You can have a look at the examples in docstring in the most recent commit. 

I will merge the `momentum_event` and `model_event` to a single event as you suggest.

Reply:
That could be a good idea. However I think it is a bit unusual to have a dict with a single key for a single EMA model (which should be a majority of usecases)...

Reply:
I see. I will update it in the next commit. Thanks a lot for the suggestions

Reply:
I updated the `attach` function as shown above. The tests are also updated. Please check them out.